exchange/bitswap/decision/strategy_prq.go

```diff
@@ -0,0 +1,225 @@
+		pQueue:   pq.New(partnerCompare),
+		rrq:      newRRQueueCustom(strategy, burst),
+	}
+}
+
+// Push
+// ----
+
+// Push adds a new peerRequestTask to the end of the list
+func (tl *strategy_prq) Push(entry *wantlist.Entry, receipt *Receipt) {
+	to := peer.ID(receipt.Peer)
+	tl.lock.Lock()
+	defer tl.lock.Unlock()
+
+	partner, ok := tl.partners[to]
+
+	if !ok {
+		partner = newActivePartner()
+		tl.pQueue.Push(partner)
+		tl.partners[to] = partner
+	}
+
+	partner.activelk.Lock()
+	defer partner.activelk.Unlock()
+
+	if partner.activeBlocks.Has(entry.Cid) {
+		return
+	}
+
+	if task, ok := tl.taskMap[taskKey(to, entry.Cid)]; ok {
+		task.Entry.Priority = entry.Priority
+		partner.taskQueue.Update(task.index)
+		return
+	}
+
+	task := &peerRequestTask{
+		Entry:   entry,
+		Target:  to,
+		created: time.Now(),
+		Done: func() {
+			tl.lock.Lock()
+			partner.TaskDone(entry.Cid)
+			tl.pQueue.Update(partner.Index())
+			tl.lock.Unlock()
+		},
+	}
+
+	partner.taskQueue.Push(task)
+	tl.taskMap[task.Key()] = task
+	partner.requests++
+	tl.pQueue.Update(partner.Index())
+
+	tl.rrq.UpdateWeight(to, receipt)
+}
+
+// Pop
+// ---
+
+// Pop 'pops' the next task to be performed. Returns nil if no task exists.
+func (tl *strategy_prq) Pop() *peerRequestTask {
+	tl.lock.Lock()
+	defer tl.lock.Unlock()
+
+	// get the next peer/task to serve
+	rrp, task := tl.nextTask()
+	if task == nil {
+		return nil
+	}
+	partner := tl.partners[rrp.id]
+
+	// start the task
+	partner.StartTask(task.Entry.Cid)
+	partner.requests--
+
+	rrp.allocation -= task.Entry.Size
+
+	if rrp.allocation == 0 {
+		// peer has reached allocation limit for this round, remove peer from queue
+		tl.rrq.Pop()
+	}
+	return task
+}
+
+// nextTask() uses the `RRQueue` and peer `taskQueue`s to determine the next
+// request to serve
+func (tl *strategy_prq) nextTask() (rrp *RRPeer, task *peerRequestTask) {
+	if tl.pQueue.Len() == 0 {
+		return nil, nil
+	}
+
+	if tl.rrq.NumPeers() == 0 {
+		// may have finished last RR round, reallocate requests to peers
+		tl.rrq.InitRound()
+		if tl.rrq.NumPeers() == 0 {
+			// if allocations still empty, there are no requests to serve
+			return nil, nil
+		}
+	}
+
+	// figure out which peer should be served next
+	for tl.rrq.NumPeers() > 0 {
+		rrp = tl.rrq.Head()
+
+		task := tl.partnerNextTask(tl.partners[rrp.id])
+		// nil task means this peer has no valid tasks to be served at the moment
+		if task == nil {
+			tl.rrq.Pop()
+			continue
+		}
+
+		// check whether |task| exceeds peer's round-robin allocation
+		if task.Entry.Size > rrp.allocation {
+			tl.partners[rrp.id].taskQueue.Push(task)
+			tl.rrq.Pop()
+			continue
+		}
+
+		return rrp, task
+	}
+	return nil, nil
+}
+
+// get first non-trash task
+func (tl *strategy_prq) partnerNextTask(partner *activePartner) *peerRequestTask {
+	for partner.taskQueue.Len() > 0 {
+		task := partner.taskQueue.Pop().(*peerRequestTask)
+		// delete task if it's trash
+		if task.trash {
+			task = nil
+			continue
```

Comment:
Setting the task to nil is not really needed. Inverting the logic would make it a bit simpler:
```go
if !task.trash {
  return task
}
```

Reply:
Hm, good point. I copied this part from the way it was done in the old PRQ, not sure why it was that way.